### DISCUSSION TITLE
How do I call the Python id function from Rust?

### DISCUSSION THREAD

**Discussion by LilyFirefly** [PyO3/pyo3#2878] [Questions] [answered]
In my specific case, I have an instance of `PyFrame`. I want to get it's `id` for use as a `HashMap` key.

**Comment by birkenfeld** [comment, marked as answer]
Since the `id` is simply the object's memory address, you can use `obj.as_ptr()` from the `AsPyPointer` trait.

Otherwise, you can get at `id` (or other built-in functions) using `py.eval("id", None, None)`.

**Reply by LilyFirefly** [reply]
Thanks!

With a bit more fiddling I got this to work:

```rust
use pyo3::AsPyPointer;

            let pyframe_id = pyframe.as_ptr() as i32;
```

**Reply by adamreichold** [reply]
Note that `i32` is not necessarily pointer size, if you just need a hash table key the `*mut PyObject` [will do just fine](https://doc.rust-lang.org/stable/std/hash/trait.Hash.html#impl-Hash-for-*mut%20T) without narrowing it to a `i32`.

**Reply by LilyFirefly** [reply]
Hmm. I'm not quite sure what that would look like.

Currently I have something like:

```rust
#[pyclass]
struct MyStruct {
    _frame_ids: HashMap<i32, String>,
}

impl MyStruct {
    fn process_frame(&self, frame: PyObject) -> Result<(), PyErr> {
        let pyframe = frame.cast_as::<PyFrame>(py)?;
        let pyframe_id = pyframe.as_ptr() as i32;
        let frame_ulid = Ulid::new();
        let frame_id = format!("frm_{}", frame_ulid.to_string());
        self._frame_ids.insert(pyframe_id, frame_id);
    }
}

**Reply by adamreichold** [reply]
I have not compiled it, but I think the following should work:

```rust
#[pyclass]
struct MyStruct {
    _frame_ids: HashMap<*mut PyObject, String>,
}

impl MyStruct {
    fn process_frame(&self, frame: PyObject) -> Result<(), PyErr> {
        let pyframe = frame.cast_as::<PyFrame>(py)?;
        let pyframe_id = pyframe.as_ptr();
        let frame_ulid = Ulid::new();
        let frame_id = format!("frm_{}", frame_ulid.to_string());
        self._frame_ids.insert(pyframe_id, frame_id);
    }
}
```

**Reply by LilyFirefly** [reply]
That doesn't play nicely with `#[pyclass]`:

```rust
error[E0277]: `*mut Py<pyo3::PyAny>` cannot be sent between threads safely                                       
    --> src/lib.rs:654:5
     |
654  |     #[pyclass]
     |     ^^^^^^^^^^ `*mut Py<pyo3::PyAny>` cannot be sent between threads safely
     |
     = help: within `(*mut Py<pyo3::PyAny>, std::string::String)`, the trait `Send` is not implemented for `*mut Py<pyo3::PyAny>`
     = help: the trait `Send` is implemented for `Py<T>`
     = note: required because it appears within the type `(*mut Py<pyo3::PyAny>, std::string::String)`
     = note: required for `hashbrown::raw::RawTable<(*mut Py<pyo3::PyAny>, std::string::String)>` to implement `Send`
     = note: required because it appears within the type `hashbrown::map::HashMap<*mut Py<pyo3::PyAny>, std::string::String, RandomState>`
     = note: required because it appears within the type `HashMap<*mut Py<pyo3::PyAny>, std::string::String>`
note: required because it appears within the type `MyStruct`
    --> src/lib.rs:656:12
```

**Reply by adamreichold** [reply]
Ah, I see. Raw pointers are not `Send` by default but that is rather a lint than a hard error (because you still need `unsafe` if you were to dereference them).

I think there are multiple options:
* Use a newtype `#[derive(PartialEq, Eq, Hash)] struct FrameKey(*mut PyObject);` and manually `unsafe impl Send for FrameKey {}`.
* Use `usize` instead of `i32` which should always suffice to store [the address part of a pointer](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.addr).
* Use `#[pyclass(unsendable)]` to avoid the `Send` requirement and you can expect the Python side to honour this.

Personally, I would probably go for `usize` in this case. (I am not sure if the additional capability bits on e.g. CHERI should actually be part of this key.)

**Reply by mejrs** [reply]
The first option is what I'd recommend.

**Reply by LilyFirefly** [reply]
I'm going with `usize` since it matches the python implementation and I understand it!

**Reply by adamreichold** [reply]
> The first option is what I'd recommend.

If we are giving conflicting advice, I guess we should add rationales so that future readers can understand why we recommended what we did. @mejrs I'll give it a try for both options but please correct me if I am wrong.

Using `usize` is close to your existing code and is simple. The possibly loss of information w.r.t. extra pointers is unlikely to be practically relevant. On downside, this is somewhat type blind and strictly speaking unnecessary as `*mut PyObject` works using the right incantations.

Using a `FrameKey` newtype is guaranteed to not lose any information and since raw pointers not being `Send` is only a lint, asserting that it actually is safe in the specific circumstances is straight forward. On the other hand, writing this is quite a bit of boiler plate and it does add an `unsafe` block even though there are fully safe alternatives (as we never want to dereference these pointers).
